python/ql/src/experimental/Security/CWE-348/ClientSuppliedIpUsedInSecurityCheckLib.qll

```diff
@@ -0,0 +1,147 @@
+private import python
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.RemoteFlowSources
+
+/**
+ * A data flow source of the client ip obtained according to the remote endpoint identifier specified
+ * (`X-Forwarded-For`, `X-Real-IP`, `Proxy-Client-IP`, etc.) in the header.
+ *
+ * For example: `request.headers.get("X-Forwarded-For")`.
+ */
+abstract class ClientSuppliedIpUsedInSecurityCheck extends DataFlow::CallCfgNode { }
+
+private class FlaskClientSuppliedIpUsedInSecurityCheck extends ClientSuppliedIpUsedInSecurityCheck {
+  FlaskClientSuppliedIpUsedInSecurityCheck() {
+    this =
+      API::moduleImport("flask")
+          .getMember("request")
+          .getMember("headers")
+          .getMember(["get", "get_all", "getlist"])
+          .getACall() and
+    this.getArg(0).asCfgNode().getNode().(StrConst).getText().toLowerCase() =
+      clientIpParameterName()
+  }
+}
+
+private class DjangoClientSuppliedIpUsedInSecurityCheck extends ClientSuppliedIpUsedInSecurityCheck {
+  DjangoClientSuppliedIpUsedInSecurityCheck() {
+    exists(RemoteFlowSource rfs, DataFlow::LocalSourceNode lsn |
+      rfs.getSourceType() = "django.http.request.HttpRequest" and rfs.asCfgNode() = lsn.asCfgNode()
+    |
+      lsn.flowsTo(DataFlow::exprNode(this.getFunction()
+              .asExpr()
+              .(Attribute)
+              .getObject()
+              .(Attribute)
+              .getObject())) and
+      this.getFunction().asExpr().(Attribute).getName() = "get" and
+      this.getFunction().asExpr().(Attribute).getObject().(Attribute).getName() in [
+          "headers", "META"
+        ] and
+      this.getArg(0).asCfgNode().getNode().(StrConst).getText().toLowerCase() =
+        clientIpParameterName()
+    )
+  }
+}
+
+private class TornadoClientSuppliedIpUsedInSecurityCheck extends ClientSuppliedIpUsedInSecurityCheck {
+  TornadoClientSuppliedIpUsedInSecurityCheck() {
+    exists(RemoteFlowSource rfs, DataFlow::LocalSourceNode lsn |
+      rfs.getSourceType() = "tornado.web.RequestHandler" and rfs.asCfgNode() = lsn.asCfgNode()
+    |
+      lsn.flowsTo(DataFlow::exprNode(this.getFunction()
+              .asExpr()
+              .(Attribute)
+              .getObject()
+              .(Attribute)
+              .getObject()
+              .(Attribute)
+              .getObject())) and
+      this.getFunction().asExpr().(Attribute).getName() in ["get", "get_list"] and
+      this.getFunction().asExpr().(Attribute).getObject().(Attribute).getName() = "headers" and
+      this.getArg(0).asCfgNode().getNode().(StrConst).getText().toLowerCase() =
+        clientIpParameterName()
+    )
+  }
```

Comment:
I suggest not quantifying over the local source node, since it will just be `rfs`. Instead I suggest adding a shorthand for `this.getFunction().(Attribute)`. And then to use `DataFlow::AttrRead` instead of `Attribute` as it operates on a higher level. Finally, I would comment the long chain of casts.
```suggestion
  TornadoClientSuppliedIpUsedInSecurityCheck() {
    exists(RemoteFlowSource rfs, DataFlow::AttrRead get |
      rfs.getSourceType() = "tornado.web.RequestHandler" and
      this.getFunction() = get
    |
      // `get` is a call to `rfs`.request.headers.get
      // `rfs`.request.headers
      get.getObject()
          .(DataFlow::AttrRead)
          // `rfs`.request
          .getObject()
          .(DataFlow::AttrRead)
          // `rfs`
          .getObject()
          .getALocalSource() = rfs and
      get.getAttributeName() in ["get", "get_list"] and
      get.getObject().(DataFlow::AttrRead).getAttributeName() = "headers" and
      this.getArg(0).asExpr().(StrConst).getText().toLowerCase() = clientIpParameterName()
    )
  }
```
A similar rewrite could be made to the one for Flask.